libraries/fc/src/network/http/http_client.cpp

```diff
@@ -0,0 +1,365 @@
+         _connections.erase(conn_itr);
+         return true;
+      } else {
+         return false;
+      }
+   }
+
+   connection_map::iterator get_connection( const url& dest, const deadline_type& deadline ) {
+      auto key = url_to_host_key(dest);
+      auto conn_itr = _connections.find(key);
+      if (conn_itr == _connections.end() || check_closed(conn_itr)) {
+         return create_connection(dest, deadline);
+      } else {
+         return conn_itr;
+      }
+   }
+
+   struct write_request_visitor : visitor<error_code> {
+      write_request_visitor(http_client_impl* that, http::request<http::string_body>& req, const deadline_type& deadline)
+      :that(that)
+      ,req(req)
+      ,deadline(deadline)
+      {}
+
+      template<typename S>
+      error_code operator() ( S& stream ) const {
+         return that->sync_write_with_timeout(*stream, req, deadline);
+      }
+
+      http_client_impl*                 that;
+      http::request<http::string_body>& req;
+      const deadline_type&              deadline;
+   };
+
+   struct read_response_visitor : visitor<error_code> {
+      read_response_visitor(http_client_impl* that, boost::beast::flat_buffer& buffer, http::response<http::string_body>& res, const deadline_type& deadline)
+      :that(that)
+      ,buffer(buffer)
+      ,res(res)
+      ,deadline(deadline)
+      {}
+
+      template<typename S>
+      error_code operator() ( S& stream ) const {
+         return that->sync_read_with_timeout(*stream, buffer, res, deadline);
+      }
+
+      http_client_impl*                  that;
+      boost::beast::flat_buffer&         buffer;
+      http::response<http::string_body>& res;
+      const deadline_type&               deadline;
+   };
+
+   variant post_sync(const url& dest, const variant& payload, const fc::time_point& _deadline) {
+      static const deadline_type epoch(boost::gregorian::date(1970, 1, 1));
+      auto deadline = epoch + boost::posix_time::microseconds(_deadline.time_since_epoch().count());
+      FC_ASSERT(dest.host(), "No host set on URL");
+
+      string path = dest.path() ? dest.path()->generic_string() : "/";
+      if (dest.query()) {
+         path = path + "?" + *dest.query();
+      }
+
+      http::request<http::string_body> req{http::verb::post, path, 11};
+      req.set(http::field::host, *dest.host());
+      req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+      req.set(http::field::content_type, "application/json");
+      req.keep_alive(true);
+      req.body() = json::to_string(payload);
+      req.prepare_payload();
+
+      auto conn_iter = get_connection(dest, deadline);
+      auto eraser = make_scoped_exit([this, &conn_iter](){
+         _connections.erase(conn_iter);
+      });
+
+      // Send the HTTP request to the remote host
+      error_code ec = conn_iter->second.visit(write_request_visitor(this, req, deadline));
+      FC_ASSERT(!ec, "Failed to send request: ${message}", ("message",ec.message()));
```

Comment:
@wanderingbort won't this assert interrupt the program ? maybe it's a bit harsh if `signature_provider` is not yet ready to sign.